yateto/ast/cost.py

```diff
@@ -71,14 +72,50 @@ def estimate_Product(self, node):
     self._cache[node] = bb
 
     return bb.size()
-  
+
   def estimate_IndexSum(self, node):
     tbb = self._cache[node.term()]
     pos = node.term().indices.find(str(node.sumIndex()))
     bb = BoundingBox([r for i,r in enumerate(tbb) if i != pos])
     self._cache[node] = bb
     return tbb.size() - bb.size()
 
+
+class GpuBoundingBoxCostEstimator(BoundingBoxCostEstimator):
+  """Estimates num. of hardware flops for a tensor operation per GPU thread.
+  Therefore, results of BoundingBoxCostEstimator are divided by a size
+  of the first dimension of lhs because this dimension is fully parallelized.
```

Comment:
Why? Even if it is parallelized, you still have to do the computation.

Reply:
![example](https://user-images.githubusercontent.com/19637079/115680058-82fa8000-a353-11eb-94c7-eca5711443e8.png)

Let's consider the example above and how it is implemented in ChainForge within a *single* GPU kernel. I need to launch a kernel with 56 threads (in fact 64 but let's skip it) because I perform matrix multiply as a sum of parallel outer products. It is worth noting that I cannot change the total number of threads inside of a kernel at run-time. If I need less threads per operation I will have to mask some of them using a simple if-statement 
```
  if (threadIdx.x < 20) {
    ...
  }
```
The rest 36 threads are going to be idle but sill waste resources i.e. registers. The problem is that a GPU has to decide how many blocks can run in a Streaming Multiprocessor (SM). Nvcc compiler computes num. required register per thread for a given kernel. Run-time system takes resources required to launch a block (registers per thread, shr. mem. per block, and required num threads per block) and SM hardware limits (max num. registers per SM, max shr. mem per SM, max num. resident threads per block) and computes how many blocks can be scheduled to an SM at run-time.  The goal is to maximize the number of resident threads per SM in order to hide global memory latency. 

It is not clear whether nvcc compiler takes into account masking or not (probably not because it estimates num. register per thread). In any case, idle threads prevents other blocks to get scheduled to the same SM.

Let's compare two implementations:

1. `(A * ((B * C) * D))`
2. `(A * (B * (C * D)))`

Num. flop per operation (1. wins)

 1. `20 x 56 x 9 + 20 x 9   x 9 + 56 x 20 x 9`
 2. `56 x 9   x 9 + 20 x 56 x 9 + 56 x 20 x 9`

Num. flop per active GPU thread. (draw)
  1. `56 x 9 + 9   x 9 + 20 x 9`
  2. ` 9  x 9 + 56 x 9 + 20 x 9`

Another aspect which is worth considering is the amount of loads from global memory to shared one because it 
 results in latency. Considering a single matrix multiply, 1) the second operand is always loaded from global to shared memory (unless it has already been loaded in a previous operation by chainforge) 2) temp. result is moved from registers to shr. memory after each operation (never goes to global memory unless the very last operation when values from registers are moved to global memory) 3) the first operand always resides in global memory (unless it needs to be transposed)

Num. glob-to-shr memory loads (2. wins)

 1. `56 x 9 + 9 x 9`
 2. `9 x 9`

Both variants lead us to the same workload per GPU thread but the second one results in less latency. The first one results in less number of flop, however, it doesn't utilize well compute resources allocated for this fused operation. The last sentence is only relevant for GPU fused GEMM kernels. For CPUs, your implementation is correct and makes much more sense.

I came up to the aforementioned thoughts by experience while I was doing some benchmarking with ChainForge. I was surprised at the beginning but the second variant runs faster. 

Answering your question below, I needed to include both things in the cost estimator to come up to a correct ordering: 1) a workload per GPU thread and 2) some penalty for loading too much to shared memory